Remote state publication (#7364)
opensearch-project/documentation-website: _tuning-your-cluster/availability-and-recovery/remote-store/remote-cluster-state.md

Setting | Default | Description
:--- | :--- | :---
`cluster.remote_store.state.index_metadata.upload_timeout` | 20s | Deprecated. Use `cluster.remote_store.state.global_metadata.upload_timeout` instead.
`cluster.remote_store.state.global_metadata.upload_timeout` | 20s | The amount of time to wait for cluster state upload to complete.
`cluster.remote_store.state.metadata_manifest.upload_timeout` | 20s | The amount of time to wait for the manifest file upload to complete. The manifest file contains the details of each of the files uploaded for a single cluster state, both index metadata files and global metadata files.
`cluster.remote_store.state.cleanup_interval` | 300s | The interval for remote state clean-up asynchronous task to run. This task deletes the old remote state files. 

## Limitations

The remote cluster state functionality has the following limitations:
- Unsafe bootstrap scripts cannot be run when the remote cluster state is enabled. When a majority of cluster-manager nodes are lost and the cluster goes down, the user needs to replace any remaining cluster manager nodes and reseed the nodes in order to bootstrap a new cluster.

## Remote cluster state publication

> The cluster manager node processes updates to the cluster state. It then publishes the updated cluster state  over the local transport layer to all of the follower nodes. With the `remote_store.publication` feature enabled, the cluster state is backed up to the remote store with every state update. The follower nodes can then fetch the state from the remote store directly which reduces the overhead on the cluster manager node for publication. 

Reviewer: ```suggestion
The cluster manager node processes updates to the cluster state. It then publishes the updated cluster state over the local transport layer to all of the follower nodes. With the `remote_store.publication` feature enabled, the cluster state is backed up to the remote store with every state update. The follower nodes can then fetch the state from the remote store directly which reduces the overhead on the cluster manager node for publication. 
```

:--- | :--- | :---
`cluster.remote_store.state.index_metadata.upload_timeout` | 20s | Deprecated. Use `cluster.remote_store.state.global_metadata.upload_timeout` instead.
`cluster.remote_store.state.global_metadata.upload_timeout` | 20s | The amount of time to wait for the cluster state upload to complete.
`cluster.remote_store.state.metadata_manifest.upload_timeout` | 20s | The amount of time to wait for the manifest file upload to complete. The manifest file contains the details of each of the files uploaded for a single cluster state, both index metadata files and global metadata files.
`cluster.remote_store.state.cleanup_interval` | 300s | The interval for the remote state clean-up asynchronous task to run. This task deletes any old remote state files. 

## Limitations

The remote cluster state functionality has the following limitations:
- Unsafe bootstrap scripts cannot be run when the remote cluster state is enabled. When a majority of cluster-manager nodes are lost and the cluster goes down, the user needs to replace any remaining cluster manager nodes and reseed the nodes in order to bootstrap a new cluster.

## Remote cluster state publication

> The cluster manager node processes updates to the cluster state. It then publishes the updated cluster state over the local transport layer to all of the follower nodes. With the `remote_store.publication` feature enabled, the cluster state is backed up to the remote store with every state update. The follower nodes can then fetch the state from the remote store directly which reduces the overhead on the cluster manager node for publication. 

Reviewer: Through would be more accurate.

To enable the feature flag for the `remote_store.publication` feature follow the steps in [experimental feature flag documentation]({{site.url}}{{site.baseurl}}/install-and-configure/configuring-opensearch/experimental/).

Enabling the setting does not change the publication flow and follower nodes will not send acknowledgements back to the cluster manager 
until they download the updated cluster state from the remote store.

> Enabling the remote cluster state feature is mandatory for remote publication to work. The routing tables repository settings contains the shard allocation details for each index in the cluster state. You can configure the remote table repository by using following settings:

Reviewer: Routing tables isn't the name of a repository. Rather, its a set of rules used that determines where certain data packets are sent [within the network](https://www.techtarget.com/searchnetworking/definition/routing-table). In this case, the routing table settings set the shard allocations for the remote cluster. However, in order for a remote cluster to work, there needs to be a repository, such as Blob Storage or S3.

I think removing the word "repository" might make the paragraph easier to understand.

> Enabling the remote cluster state feature is mandatory for remote publication to work. The routing tables repository settings contains the shard allocation details for each index in the cluster state. You can configure the remote table repository by using following settings:

Reviewer: ```suggestion
Enabling the remote cluster state feature is mandatory for remote publication to work. To modify the remote publication behavior, the following routing table repository settings contains the shard allocation details for each index in the remote cluster state:
```

Reviewer 2: @Naarcha-AWS Edited the suggestion

```yml
# Remote routing table repository settings
node.attr.remote_store.routing_table.repository: my-remote-routing-table-repo
node.attr.remote_store.repository.my-remote-routing-table-repo.type: s3
node.attr.remote_store.repository.my-remote-routing-table-repo.settings.bucket: <Bucket Name 3>
node.attr.remote_store.repository.my-remote-routing-table-repo.settings.region: <Bucket region>
```

You do not have to use different remote store repositories for state and routing, since both state and routing can use the same repository settings.

To configure remote publication, use the following cluster settings:

Setting | Default | Description
:--- | :--- | :---
`cluster.remote_store.state.read_timeout` | 20s | The amount of time to wait for remote state download to complete on the follower node.

> `cluster.remote_store.routing_table.path_type` | HASHED_PREFIX | Path type to be used for creating index routing path in blob store. Valid values are "FIXED", "HASHED_PREFIX", "HASHED_INFIX"

Reviewer 2: ```suggestion
`cluster.remote_store.routing_table.path_type` | HASHED_PREFIX | The path type to be used for creating index routing path in blob store. Valid values are "FIXED", "HASHED_PREFIX", and "HASHED_INFIX".
```

> `cluster.remote_store.routing_table.path_type` | HASHED_PREFIX | Path type to be used for creating index routing path in blob store. Valid values are "FIXED", "HASHED_PREFIX", "HASHED_INFIX"

Reviewer 2: It looks like there are a couple of articles missing in the description. Should it be "The path type to be used for creating an index routing path in the blob store"?

> `cluster.remote_store.routing_table.path_hash_algo` | FNV_1A_BASE64 | Algorithm to be used for constructing prefix or infix of blob store path. This setting comes into effect into if cluster.remote_store.routing_table.path_type is "hashed_prefix" or "hashed_infix". Valid values of algorithm are "FNV_1A_BASE64" or "FNV_1A_COMPOSITE_1"

Reviewer 2: ```suggestion
`cluster.remote_store.routing_table.path_hash_algo` | FNV_1A_BASE64 | The algorithm to be used for constructing prefix or infix of blob store path. This setting is applied if `cluster.remote_store.routing_table.path_type` is "hashed_prefix" or "hashed_infix". Valid algorithm values are "FNV_1A_BASE64" and "FNV_1A_COMPOSITE_1".
```

> `cluster.remote_store.routing_table.path_hash_algo` | FNV_1A_BASE64 | Algorithm to be used for constructing prefix or infix of blob store path. This setting comes into effect into if cluster.remote_store.routing_table.path_type is "hashed_prefix" or "hashed_infix". Valid values of algorithm are "FNV_1A_BASE64" or "FNV_1A_COMPOSITE_1"

Reviewer 2: Same comment re: articles. Should it be "The algorithm to be used for constructing a prefix or infix of the blob store path"? Also, here and in the preceding row, should the text in quotation marks actually be in code font?

Reviewer: Yes. Done and fixed.

Setting | Default | Description
:--- | :--- | :---
`cluster.remote_store.state.index_metadata.upload_timeout` | 20s | Deprecated. Use `cluster.remote_store.state.global_metadata.upload_timeout` instead.
`cluster.remote_store.state.global_metadata.upload_timeout` | 20s | The amount of time to wait for cluster state upload to complete.
`cluster.remote_store.state.metadata_manifest.upload_timeout` | 20s | The amount of time to wait for the manifest file upload to complete. The manifest file contains the details of each of the files uploaded for a single cluster state, both index metadata files and global metadata files.
`cluster.remote_store.state.cleanup_interval` | 300s | The interval for remote state clean-up async task to run. This task deletes the old remote state files. 

## Limitations

The remote cluster state functionality has the following limitations:
- Unsafe bootstrap scripts cannot be run when the remote cluster state is enabled. When a majority of cluster-manager nodes are lost and the cluster goes down, the user needs to replace any remaining cluster manager nodes and reseed the nodes in order to bootstrap a new cluster.

## Remote Cluster State Publication

> The cluster state published to remote-backed storage can be used for publication. Currently, the active cluster manager

Reviewer 3: The cluster manager node processes the updates to cluster state. It then, publishes the updated cluster state over the local transport layer to all the follower nodes. With remote cluster state enabled, cluster state is backed to remote store with every state update. The follower nodes can fetch the state from remote store directly and reducing the overhead on the cluster manager node for publication. This can be done by enabling the experimental remote publication feature. 
Enable the feature flag for `remote_store.publication` feature by following the [experiment feature flag documentation]({{site.url}}{{site.baseurl}}/install-and-configure/configuring-opensearch/experimental/). This doesn't change the publication flow and follower nodes will not send acknowledgement back to cluster manager until they download the updated cluster state from remote store and proceed as expected in current flow. Please note enabling remote cluster state is mandatory for remote publication to work. 
Also, RoutingTable which contains the shard allocation details for each index in the cluster state requires setting up the remote blobstore repository. It can be configured as below:
